### ISSUE TITLE
Multiple UPE - Stripe Link compatibility

### ISSUE THREAD

**Issue by harriswong** [Automattic/woocommerce-payments#4720]
### Describe the issue
This is part of the Multiple UPE project (https://github.com/Automattic/woocommerce-payments/pull/3968). 

The multiple UPE POC uses classic element for credit card checkout, but Stripe Link requires UPE. Does this mean Stripe Link no longer work with multiple UPE? If so, how should we handle this? 

### Accepted criteria
- Find out if the multiple POC branch works with Stripe Link 
- If not, please spearheads a conversation/solution with the team(s).

**Comment by Mayisha** [comment]
[Stripe Link](https://woocommerce.com/document/payments/woocommerce-payments-stripe-link/) seems very similar to WooPay in general. When Stripe Link is enabled-

- On the checkout page the email field is rendered on top of Billing Address (same as WooPay)
- When a user inputs an email address that does not exist in Stripe Link, a checkbox appears below the credit card option. (very similar experience as WooPay)
![new user](https://user-images.githubusercontent.com/33387139/190491060-4dc77380-bb68-4310-bc9e-12e4df534d49.png)

- On place order, a Link user is created (same experience as WooPay)
- When a user inputs an existing Link email address, an OTP Modal appears. (same experience as WooPay)
![otp](https://user-images.githubusercontent.com/33387139/190491307-531c7b4d-5ca0-4616-ac79-0e7cf12d1234.png)

- On successful OTP verification, another modal appears with the account data and a button to autofill the current merchant checkout page with the selected Link account data (this step is different from WooPay as the user always stays on the merchant's checkout page)
![account](https://user-images.githubusercontent.com/33387139/190491353-eb6d201b-564a-422b-93d3-eb0cfa9ba371.png)

- The OTP modal appears when the user types in an existing email. Otherwise, for prefilled email field the input box shows a button that initiates the OTP modal. (WooPay does not have this extra button)
![link button](https://user-images.githubusercontent.com/33387139/190491730-0db1637c-e3ad-4ae5-abc8-bd3af366c518.png)

**Comment by harriswong** [comment]
Thanks @Mayisha for the screenshot and notes!

🤔  If we are using the same email address, how are we going to switch between WooPay or Link?

**Comment by Mayisha** [comment]
> Find out if the multiple POC branch works with Stripe Link

### Existing Link customer flow 
✅ When a user inputs an existing Link email address, an OTP Modal appears as expected.
❌ The Link button also appears on the right corner of the input box when it should not be there.
❌ In the payment methods list there is an option for Link. For an existing user, the Link method should not be there (also not as a separate method like this)
<img src="https://user-images.githubusercontent.com/33387139/190494427-822dc529-0af8-42ef-b667-833c1c6a29b2.png" height="50%" width="50%" />

❌ When I click `Autofill Checkout` some fields are not filled (first name, last name, phone number)

### New Link customer flow
❌ In the payment methods list there is a separate option for Link instead of the checkbox below credit card option. If I expand the Link option, it's empty.
<img src="https://user-images.githubusercontent.com/33387139/190494971-4999e8d0-5ad6-4cf3-b591-14ecfe1a398b.png" height="50%" width="50%" />
❌ There is no checkbox to allow creating a Link user.

**Comment by Mayisha** [comment]
> If we are using the same email address, how are we going to switch between WooPay or Link?

@harriswong this is a very important question and we need some decisions on how we are going to allow merchants to enable Link and WooPay. As the existing user check happens when the user inputs a valid email field and the OTP modal appears for an existing user, for an email that has both Link and WooPay accounts it gets messy :exploding_head: 

![link vs woopay](https://user-images.githubusercontent.com/33387139/190496769-4e14460f-000a-4db3-a99b-5bf53d0288af.png)

**Comment by frosso** [comment]
Since Link is only compatible with the UPE (and not with Stripe Elements), what are people's thoughts on showing Link as a separate gateway entry on the checkout page (when WooPay is enabled)?

Similar to this:
<img src="https://user-images.githubusercontent.com/33387139/190494427-822dc529-0af8-42ef-b667-833c1c6a29b2.png" height="50%" width="50%" />

Granted, we would still need to resolve [some compatibility issues](https://github.com/Automattic/woocommerce-payments/issues/4720#issuecomment-1248546228), like @Mayisha pointed out. Maybe in this case we would allow Stripe to show the Link login form only when Link is "expanded" on the checkout page, and force Link to be the last gateway items in the list of gateways? (so that a customer is "forced" to click on it to expand it)

Additionally, we would need to make some decisions on what to do when:
- Merchant enables both Link & WooPay
- Customer goes to checkout page
- Customer expands "Link" gateway entry
- Customer enters email address
In this case, we might have to honor the customer's desires and _not_ show the WooPay login window (allowing them to see only the Link login window)

**Comment by Mayisha** [comment]
>  Since Link is only compatible with the UPE (and not with Stripe Elements), what are people's thoughts on showing Link as a separate gateway entry on the checkout page (when WooPay is enabled)?

@frosso do you mean to show this for the Link's existing user flow or the new user flow? If we are changing Link's [current experience](https://github.com/Automattic/woocommerce-payments/issues/4720#issuecomment-1248520953) we should probably discuss with the people owning the Link feature as well. Another question that comes to my mind is, is the UI/UX of Link done by us? Or do we have any contract with Stripe that we are supposed to offer their feature in this and this way? :thinking: 

For new users, there is a checkbox below the credit card and the payment method is `card` at that time. If we make Link a separate option, then `card` and `link` are 2 different payment methods there and the user can choose only one. But again Link account is not there yet, so can not possibly pay with Link.

> Maybe in this case we would allow Stripe to show the Link login form only when Link is "expanded" on the checkout page, and force Link to be the last gateway item in the list of gateways?

For existing users, after OTP verification and autofill with Link, the credit card input box is replaced with the card details from Link.
<img src="https://user-images.githubusercontent.com/33387139/190581038-31e89054-7fd8-4dde-bef0-0ffe96d02b2d.png" width="50%" height="50%" />

If we make Link a separate method here then there will be the traditional credit card option with the cc input field and at the bottom the Link option. We have to then render our own components which will trigger the Link login/ OTP modal (I don't know the technical implementation details yet)

For both cases, it looks like a big refactor in user experience. And also we have many lingering questions that you added at the end that calls for a product decision and design decision. :eyes:

**Comment by harriswong** [comment]
I like the idea of splitting "Credit card/ debit card" and "Link" into 2 separate payment methods. In this case, woopay can attach onto "Credit card/ debit card" and Link is its own "credit card". 

👋 @rogermattic Looping you in regarding some changes around how we should display Link, "Credit card/ debit card" and multiple UPE here.

**Comment by rogermattic** [comment]
👋 @harriswong  thanks for pinging me here! 

Thank you for the great exploration on this @Mayisha and for all your inputs @frosso !

This is a tough one, cause these are conflicting gateways and I assume that if both are enabled, then we want to prioritize WooPay. From the first glance on it, I also like splitting "Credit card/ debit card" and "Link" into 2 separate payment methods, as it seems the least confusing. Having said that, I need to test it further. Am I understanding this correctly that this usecase covers a new user experience?)

The email issue (for an existing user) seems to be even more complex, but again.. I need to test it further to make sure I have a deeper understanding. I don't think we ever want to show two competing modals as (like @Mayisha said) that gets really messy and confusing. I think the solution I think I may consider is offering the user an informed choice like "your email address is associated with WooPay & Link, which one do you want to checkout with" or something along these lines.. so that they can choose which gateway they would like to go for and we can streamline the experience and simplify it. 

With help of @mdmoore I'll try to setup a test environment so that I can have a good look into this and let you know more thoughts asap.

**Comment by pierorocca** [comment]
From Britni 

> [1:22](https://a8c.slack.com/archives/D03BYV1JTTR/p1664389375405689)
Regardless, we've fulfilled the milestone by having Link available as a payment method. That was the gist of our commitment.

It would be preferable to have Link as a method within the UPE stack than have it compete or have it muddle up the WooPay UI/UX. Another alternative that could be explored is to have Link as an express checkout button only with no OTP trigger from the email form field.

**Comment by rogermattic** [comment]
Following up on my previous comment, here are some brief ideas on what we could do (as an alternative to the ideas discussed above):

Link could be available as one of the express checkout buttons:
![d](https://user-images.githubusercontent.com/94967682/194124491-d1b55794-5f11-4271-a1d8-7f0565463a84.png)


If we decide to stick with a modal, I thought something like this could work (WIP). In a nutshell, we could keep a single modal and introduce an alternative button within the modal:
![auth_resting](https://user-images.githubusercontent.com/94967682/194121854-d1033f61-6981-4860-992d-42d251bcb0f0.png)

I'm also considering following changes in Settings – currently exploring a concept of a `default` method. 
![Settings _ Payments _ WooCommerce Payments _ Payment methods _ None](https://user-images.githubusercontent.com/94967682/194127510-a2daf52e-7616-4534-92c1-1ef8e7df695d.png)


The default method would be prioritized in the email field (and the checkout modal) on the customer facing side. 
 It would obviously require some explanation to the user (in the UI). 
 
 Let me know your thoughts!

**Comment by Mayisha** [comment]
@rogermattic the designs look great and with WooPay being visible in form of a button makes things straight forward from the user's perspective as well. :tada: 

About the `default` method setting I feel like a `default` payment method should be more of a shopper's choice than the merchant's choice. Though I am allowed to switch, one shop always opening WooPay for me, another always opening Link for me would feel a bit weird to me as shopper. :thinking:  This is completely my personal feeling though :sweat_smile:

**Comment by FangedParakeet** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Heisenberg-63586f2ec60359001310fa6a/issues/Automattic/woocommerce-payments/4720?planning-poker) with Zenhub @mdmoore @timur27

**Comment by FangedParakeet** [comment]
@rogermattic, I beseech your blessings, if I may, for my implementation approach in #5282. Allow me to summarise the highlights from my PR.

>Stripe Link is an express checkout payment method that requires a UPE card payment element at checkout and responds to changes from the email input field. The split-UPE implementation does not use the UPE for the card element in order to maintain compatibility with WooPay and moreover WooPay also listens to changes in the email input field, which causes WooPay and Link's authentication modals to overlap each other disgracefully on the checkout.

...

>Ideally we would like to implement the following functionality.
>- Stripe Link appears as an express checkout payment button.
>- Stripe Link modal and WooPay modals do not appear at the same time.

...

>[The linkAutofillModal interface] creates the following challenges to our objectives stated above.
>- We have no way of knowing whether a user's email address is a registered Stripe Link account, until after they have successfully validated via OTP.
>- Once we "begin" the Stripe Link checkout, we have no way of controlling or restraining the OTP modal from appearing on the page. If we query for a user's registration and the user is found, the modal will open on the page.
>- If we query Stripe Link for a user's registration and the OTP modal does not open, it could be because the query is still ongoing or it could be because the user has not been found and we have no way of distinguishing between either state.

...

>As a result of these conditions, this is the compromise that I have concluded upon in this PR.
>
>Stripe Link is present as a payment request button. Currently, it will not automatically listen to changes on the email input, as WooPay does. However, if a user selects to checkout with Stripe Link, we will begin sending the email input values to Link instead and regardless of the result (since this information is not available to us), the WooPay modal will be blocked by Link. 
>
>In this scenario, selecting to checkout with Stripe Link means either the user has clicked the Stripe Link express checkout button or has selected the Link checkout gateway radio button.  
>
>So if a user has a single email address associated with both WooPay and Stripe Link, the following two paths are possible at checkout.
>
> If the user arrives at the checkout and does not actively select Link, the WooPay modal will appear and Stripe will be none the wiser of the user's presence.

![GIF of WooPay modal opening](https://cdn-std.droplr.net/files/acc_1104206/6X85k6)
_WooPay takes precedent when Link is not selected_

> However, if the user selects Link while we are still querying WooPay, we will instead switch to the Link checkout, the WooPay modal will be blocked, and Link's modal will open instead.

![GIF of Link modal opening](https://cdn-std.droplr.net/files/acc_1104206/Ev9X9G)
_Selecting Stripe Link blocks WooPay_

This will likely make more sense in action, so I have commissioned a Jurassic Ninja test site for your viewing pleasure at the following location.

**Site:** ~~https://interior-spinner.jurassic.ninja~~ `https://historic-krill.jurassic.ninja`
**Login:** `demo`/~~dKT0f2vdaNZh~~ `zoI9cAbnGXG5`

Just add an item to cart and proceed to the shortcode checkout and you should be able to interact with both the Link and WooPay checkout buttons as they vie for your attention. It has not yet been completed, but if confirmed, this similar flow and button relationship will be inherited across to the blocks checkout as well (see #5340).  Please let me know if anything in this approach greatly offends your senses or if we are safe to continue on down this path. Cheers!

Edit: Updated demo site links. ☝️

**Comment by frosso** [comment]
@elizaan36 since Magda will be AFK for a little while, would you be a good point of contact to take over this? 🤗

**Comment by elizaan36** [comment]
hey @frosso @FangedParakeet Will this also be updated in the blocks checkout?

@stephendharmadi, could you take a look and share your feedback on this approach?

**Comment by timur27** [comment]
Hey @elizaan36! Stripe Link compatibility with Blocks will be handled by a follow-up issue https://github.com/Automattic/woocommerce-payments/issues/5340. Within the current https://github.com/Automattic/woocommerce-payments/issues/4720 issue, the Blocks checkout does not work with Stripe Link.

**Comment by stephendharmadi** [comment]
Hey, I will share feedback from my understanding.

From the issue that was already described above. I assume the OTP modal will only appear if the system can send a unique code to the registered phone number from the system, which could mean the buyer's email address is already registered too. Am I right?

We have to make sure there will be only a single OTP modal appearing at the time, therefore please refer to these conditions below (with the diagram):

- **Use case 1 - When the buyer's email address is registered to WooPay.**
By default, WooPay modal will always appear first (assuming that WooPay can detect the registered email without OTP inputted).
It can only be replaced by the Stripe Link OTP modal if buyers intentionally select or click "Stripe by Link" express payment button or the Link checkout gateway radio button. If buyers closed the Stripe Link OTP modal and click on the email address field again, the WooPay modal will appear again. This how we make sure it won't be blocked by Link.

- **Use case 2 - When buyer's email address is NOT registered to WooPay.**
We could run the Stripe by Link OTP modal automatically ONLY if the email inputted isn't registered to WooPay.

> If we query Stripe Link for a user's registration and the OTP modal does not open, it could be because the query is still ongoing or it could be because the user has not been found and we have no way of distinguishing between either state.

I just checked the test store you shared @FangedParakeet and clicked the Link by Stripe button and the modal didn't appear. We have to inform buyers when the OTP modal can't be opened.

Is there a time limit we can set to ensure that the query's no longer in progress (e.g., 60 seconds)? We can show the loading state during the time. After it reached the time limit, we could inform buyers with a modal stating the user has not been found. (I will provide the mockup later).

![Diagram](https://user-images.githubusercontent.com/110893979/210967889-24f4aa3d-d5eb-4f7b-a6f5-7ea61fcf0ba6.png)

**Comment by FangedParakeet** [comment]
> I assume the OTP modal will only appear if the system can send a unique code to the registered phone number from the system, which could mean the buyer's email address is already registered too. Am I right?

Yes, you are indeed correct!

> By default, WooPay modal will always appear first (assuming that WooPay can detect the registered email without OTP inputted).

Cool, I can ensure that Link always waits for the response from WooPay's `/user/exists` endpoint before making any requests, unless the Link button is explicitly selected. For now this will have to be achieved by having one front-end component poll another on an interval, but in the future--once all gateways are able to use the UPE elements when the UPE is enabled--this will have a more synchronous, cooperative design. 

> We could run the Stripe by Link OTP modal automatically ONLY if the email inputted isn't registered to WooPay.

👍 

> If buyers closed the Stripe Link OTP modal and click on the email address field again, the WooPay modal will appear again. This how we make sure it won't be blocked by Link.

Unfortunately, one limitation of the API Stripe offers us to implement Link is that once the Link checkout is initiated we have no obvious way of knowing if the OTP modal has opened and, if it opens, we will have no way of knowing that a user has closed it.

> Is there a time limit we can set to ensure that the query's no longer in progress (e.g., 60 seconds)?

This is challenging, because we currently have no obvious way of knowing that the OTP modal has opened nor that it has been closed. If the email address is not recognised by Link, Stripe's "Save your info" checkbox will not appear; but this is an iframed element, so I'm not sure that I'm able to poll for its presence on the page. 

Furthermore, if the OTP modal does open and the user closes it, there is no event for us to learn that this has occurred. Once again this content is controlled by Stripe and encapsulated within their own iframe, where we have no jurisdiction to listen to DOM events nor to capture clicks. Consequently once we initiate a Link checkout, we cannot differentiate between a user waiting to learn whether they are registered or a registered user waiting to receive an OTP message. We are only provided with an event when the user successfully authenticates via OTP. 

We could implement a larger timeout for a user's authentication and upon the expiration of this allotted time we could return control back to WooPay. However, this timeout window would have to be somewhere in the region of 3-5 mins, as an unsuccessfully authenticated user could either be an unregistered Link user or a valid registered Link user who either closed the modal or failed to enter the OTP in time.

This also means that while I can implement a loading state for Link, I can only know for sure that it should be removed once a user successfully completes their authentication with Stripe. For example, if a user is recognised by Stripe, presented with an OTP, and awaits its receipt, we have no way of knowing that this progress has been made, so the loading state would remain. 

I'm going to double-check these statements, to confirm which elements we can and cannot poll, but sadly this updates your flowchart as follows with a few unhappy dead-ends.

![Updated Link flow](https://cdn-std.droplr.net/files/acc_1104206/uXQ6Yh)
_Sad Stripe noises_

Let me know what you think about these earth-shattering ramifications, @stephendharmadi. In the meantime, I will update #5282 to always await the response from WooPay's registration endpoint, prior to initiating anything with Link.

**Comment by stephendharmadi** [comment]
Hi @FangedParakeet , Thank you for thoroughly explaining what is possible and not possible for the proposed solutions.

> Unfortunately, one limitation of the API Stripe offers us to implement Link is that once the Link checkout is initiated we have no obvious way of knowing if the OTP modal has opened and, if it opens, we will have no way of knowing that a user has closed it.

> We could implement a larger timeout for a user's authentication and upon the expiration of this allotted time we could return control back to WooPay. However, this timeout window would have to be somewhere in the region of 3-5 mins, as an unsuccessfully authenticated user could either be an unregistered Link user or a valid registered Link user who either closed the modal or failed to enter the OTP in time.

I think the allotted time of 3-5 mins is a reasonable time to let buyers through the authentication process with Stripe. Let's implement this and revert buyers back to the WooPay OTP modal if there's an unsuccessfully authenticated user (could either be an unregistered Link user or a valid registered Link user who either closed the modal or failed to enter the OTP in time).

> This also means that while I can implement a loading state for Link, I can only know for sure that it should be removed once a user successfully completes their authentication with Stripe. For example, if a user is recognised by Stripe, presented with an OTP, and awaits its receipt, we have no way of knowing that this progress has been made, so the loading state would remain.

I still see the need to display the loading state, instead of letting buyers feel confused when nothing happens when the Link button is clicked. Could we still display the loading state with 2 conditions below:
- If buyers successfully completed the authentication, then the loading state is removed.
- If it's presented with an OTP and awaits its receipt. Could the loading state remain but hidden under the process shown? So buyers don't have to see it until the authentication is complete.

I have updated the diagram to reflect this latest discussion. Let me know what you think about this mitigation.
![Flow iteration 2](https://user-images.githubusercontent.com/110893979/211473000-2a24cba1-551e-400e-a64c-38ebe661c759.png)

**Comment by FangedParakeet** [comment]
Hey, @stephendharmadi, that flowchart looks pretty tidy to me, but I've just a few more clarifications before I can give life to your vision.

> I think the allotted time of 3-5 mins is a reasonable time to let buyers through the authentication process with Stripe.

I'd be willing to implement this five minute timeout...However, just as a caveat, I may have made a few inaccurate presumptions pertaining to Stripe's OTP process. I had assumed that Stripe's OTP SMS codes were time-based codes, similar to WooPay's codes. WooPay verification codes use a time-based hashing function that uses the current timestamp to create a unique hash for a user. Instead of storing the hash, we verify hashes by comparing them against valid hashes in an acceptable five minute time-window. Stripe appears to have much longer-lived hashes and probably doesn't use a time-based hash, [since their 2FA SMS verification codes are valid for up to thirty-minutes](https://support.stripe.com/questions/not-receiving-text-message-(sms)-verification-code-for-two-step-authentication-verification). I'm not sure that Stripe Link OTP codes are generated in the same way as their 2FA authentication codes, but since they are both six-digit codes, it's quite likely.

Let me first apologise since this was probably a needlessly long explanation, but in short the takeaways are that while WooPay codes are valid for five minutes and we assume that an OTP code will be received within this window, it's possible that it could take up to thirty minutes for a Link OTP code to be delivered. I think it's quite unlikely that this would occur nor that any sane user would actually stoically wait for thirty minutes at checkout to receive an OTP code; but the problem is that if we revert control back to WooPay, since there may be no way of knowing whether Stripe's OTP modal is current open, if the user is also registered to WooPay, we will end up presenting both OTP modals simultaneously superimposed, which is the nightmare situation we are trying to avoid in this issue! 

If Link's OTP transmission takes longer than anticipated, our most tragic nightmares might become reality and I don't see an obvious solution to this problem, other than increasing the timeout to thirty minutes, which seems to be fruitlessly lengthy.

> Let's implement this and revert buyers back to the WooPay OTP modal if there's an unsuccessfully authenticated user (could either be an unregistered Link user or a valid registered Link user who either closed the modal or failed to enter the OTP in time).

If you'll entertain it, please allow me to share the following scenario, in the case where a user is registered to both WooPay and Link.

1. User enters checkout.
2. User is registered to WooPay, so WooPay modal appears first.
3. User actually prefers Link, so closes the WooPay modal and selects Link instead.
4. Link OTP modal now appears, but the user either changes their mind or the OTP verification fails for some reason and our authentication timeout period expires.
5. Control is returned to WooPay and WooPay checks the registration status of the email address in the email input.

Should we now automatically reopen the WooPay modal for a second time, since we confirm again that the user's email address is registered to WooPay? Or should we presume that the user has declined both options and keep the WooPay modal closed unless the user specifically chooses to restart this process by clicking the WooPay express checkout button?

> I still see the need to display the loading state, instead of letting buyers feel confused when nothing happens when the Link button is clicked.

Sure, that makes perfect sense. For a rather important yet quite inept follow-up question, can you just confirm what the loading state is? 😅 Is that the spinning preloader that appears next to the email input? Just FYI, that currently only signals activity from WooPay, but I'm pretty sure I can repurpose this for Link as well, if desired. If you had some other loading state in mind, please let me know!

> If buyers successfully completed the authentication, then the loading state is removed.

Cool beans, this can be done!

> If it's presented with an OTP and awaits its receipt. Could the loading state remain but hidden under the process shown? So buyers don't have to see it until the authentication is complete.

I cannot actually tell when Stripe's OTP modal has opened, so this is a little more challenging. However, this seems like a more suitable place for a timeout to be put in place. While Link's OTP delivery may be quite extended, their email recognition appears to be relatively more rapid. After commencing the Link checkout flow, after about 30-60s I think we can safely assume that either the OTP modal has opened or the user's email address has not been recognised. In either state, I think we can safely remove the loading state at this time.

It may take more time for the OTP to actually be delivered and I cannot know for sure whether the user is still in the middle of the Link authentication process, but I think we can confidently that after 30-60s the user's email address has either been recognised or not and we can safely remove the loading state. The upper bound (60s) is probably safer, but the lower bound (30s) is probably a more rational and functional time to maintain a preloader, so let me know which you prefer and I'll implement it!

Cheers and sorry again for the ominous wall of text.

**Comment by FangedParakeet** [comment]
@stephendharmadi whenever you have a spare moment, I've created a new demo of this functionality that aspires to follow your flowchart's designs in [this comment](https://github.com/Automattic/woocommerce-payments/issues/4720#issuecomment-1376772755). Please could you give this one last look and let me know if Link's functionality satisfies you?

Both Link and WooPay will now automatically use the email field to query for registration; however, Link will always wait for WooPay to complete its queries first. If WooPay fails to match the email address to a registered user, Link will consequently try to do so. If WooPay does match a registered user, WooPay's modal will open first and Link will remain dormant. However, the user is free to close the WooPay modal and trigger a checkout with Link by clicking on the Express checkout button. 

For more details, feel free to explore the enhancements to #5282. To dive into the demo, please follow the link below.

https://increasing-mosquito.jurassic.ninja/
`demo`/`L96bdNzWWekw`

**Comment by stephendharmadi** [comment]
Hey @FangedParakeet. Sorry, I just had time to reply to this thread.

>If Link's OTP transmission takes longer than anticipated, our most tragic nightmares might become reality and I don't see an obvious solution to this problem, other than increasing the timeout to thirty minutes, which seems to be fruitlessly lengthy.

Sorry if you have explained this to me, but do we have control over the LInk by Stripe OTP? We have to consider that buyers could select Link by Stripe and then decide to use WooPay again. I would expect the scenario when buyers selected WooPay during Link's OTP transmission (which might take up to 30 minutes), or buyers leave the checkout and return back to the checkout. It would be better if we could turn off the Link's OTP transmission in these cases, and prevent it from appearing together with WooPay's OTP.

> Should we now automatically reopen the WooPay modal for a second time, since we confirm again that the user's email address is registered to WooPay? Or should we presume that the user has declined both options and keep the WooPay modal closed unless the user specifically chooses to restart this process by clicking the WooPay express checkout button?

Yes! it makes more sense to reopen the WooPay modal for users with email address registered to WooPay, if:
- Users intentionally clicked the WooPay express checkout button during the LInk's OTP transmission.
- Users left the checkout page during Link's OTP transmission, returned back, and reenter their registered email address.

> Sure, that makes perfect sense. For a rather important yet quite inept follow-up question, can you just confirm what the loading state is? 😅 Is that the spinning preloader that appears next to the email input? Just FYI, that currently only signals activity from WooPay, but I'm pretty sure I can repurpose this for Link as well, if desired. If you had some other loading state in mind, please let me know!

I would expect the same loading state as what we have in WooPay, so we can repropose the same pattern for Link as well. I don't think it's necessary to create another loadings state just for Link.

> It may take more time for the OTP to actually be delivered and I cannot know for sure whether the user is still in the middle of the Link authentication process, but I think we can confidently that after 30-60s the user's email address has either been recognised or not and we can safely remove the loading state. The upper bound (60s) is probably safer, but the lower bound (30s) is probably a more rational and functional time to maintain a preloader, so let me know which you prefer and I'll implement it!

It sounds good to me. I think we can implement the small loading state in the email address text field, to make it less obvious and won't distract the OTP modal. (example below)
<img width="347" alt="Screenshot 2023-01-24 at 16 35 57" src="https://user-images.githubusercontent.com/110893979/214258495-d9f6b207-8e63-419e-8271-8fc8b9b2c2ea.png">

**Comment by stephendharmadi** [comment]
Hey @FangedParakeet, thanks for sharing the demo. I will take a look.

> For more details, feel free to explore the enhancements to https://github.com/Automattic/woocommerce-payments/pull/5282. To dive into the demo, please follow the link below.
https://increasing-mosquito.jurassic.ninja/demo/L96bdNzWWekw

Clicked the link you shared but it shows as "not found", also could you share with me the email address registered for both WooPay and Link so that I can do a quick test on the demo?

Also, I just noticed this warning banner. I think it's not a good experience to warn users and impose them to input email to use Link by Stripe. Could we remove it? or at least, may I know the rationale behind this decision?

cc @elizaan36 
 
<img width="926" alt="Screenshot 2023-01-24 at 16 51 11" src="https://user-images.githubusercontent.com/110893979/214261184-f9ca3cc6-6a6d-437c-abd0-4f9cdf35382c.png">

**Comment by FangedParakeet** [comment]
> do we have control over the Link by Stripe OTP?

No. 😢 

> It would be better if we could turn off the Link's OTP transmission in these cases

This is sadly not possible. There are no brakes on the OTP transmission train after we initialise Link and query Stripe.

> We have to consider that buyers could select Link by Stripe and then decide to use WooPay again. I would expect the scenario when buyers selected WooPay during Link's OTP transmission (which might take up to 30 minutes), or buyers leave the checkout and return back to the checkout.

FWIW, here is how things currently work in #5282 (**TL;DR, I believe your description is already how things currently work after my latest amends**). 

Whenever an email address is entered in the email input, we will _always_ wait until WooPay has been queried before querying Link. So if Stripe is even queried at all, we know that either the user's email address is not registered to WooPay or it is registered to WooPay, but the user has closed the WooPay modal and clicked the Link express button. 

If the user is also registered to Link, now Link's OTP modal will open and the SMS transmission will begin. It is not possible to select the WooPay express button while the Link OTP modal is open, because the modal appears on top of other elements on the page. The only way to select WooPay at this point is to either close the Link OTP modal or enter a valid OTP code, authenticate with Stripe, and then close the Autofill modal. When Link's OTP modal is closed, the Link checkout cannot be restarted and the Link express button is permanently disabled; when the Autofill modal is closed, this can, on the other hand, be reopened. However, in either case, clicking the WooPay express button after either modal has been closed should restart the WooPay checkout.

> I would expect the same loading state as what we have in WooPay, so we can repurpose the same pattern for Link as well.

Cool, this was actually implemented in 47e7abd8a677675be39518efb0ac084c6f7a0e09. 👍 

> Also, I just noticed this warning banner. I think it's not a good experience to warn users and impose them to input email to use Link by Stripe. Could we remove it?

I actually removed this in 50fa3e91040405cd77c55520618e216c452dc54d after a chat with @elizaan36. 😅 👍

**Comment by stephendharmadi** [comment]
It sounds good to me! Thanks for working on it.  🚢 it~

**Comment by stephendharmadi** [comment]
Hi, based on the discussion in [this slack thread.](https://a8c.slack.com/archives/C022WMN88KG/p1674664908875189)

The one I proposed above was intended to provide a seamless checkout experience for buyers, without considering WooPay conversion. If the purpose is to gain WooPay conversion from new users, then it makes sense not to trigger the link OTP automatically even if an inputted email address IS NOT registered to WooPay and IS registered to Link. Yes, we can implement it, so Link will always wait for a button click before springing into action although an inputted email address IS registered to Link. I see no big issue here from a design perspective.

Please find the updated flowchart below.
![Flow iteration 3](https://user-images.githubusercontent.com/110893979/215041273-fbdd159b-054e-4bf0-bf8a-6e545c3d4ac7.png)

**Comment by FangedParakeet** [comment]
After a lot of discussion, [it has been decided ](p1675199840291439/1674664908.875189-slack-C022WMN88KG)that ultimately despite our best efforts, it seems like WooPay and Link cannot coexist together without being at each other's detriment, so these payment methods will now operate mutually exclusively.

WooPay can only be enabled when Link is disabled and Link can only be enabled when WooPay is disabled. We will restore the existing Link implementation to the split UPE branch instead and make sure that it is still compatible with our shortcode and blocks checkouts: this may involve adding a new gateway for Link. Nonetheless when active Link will listen to the email field and trigger automatically as usual, unfettered by WooPay.

We will need to add the following changes to the settings pages, wherein when WooPay is enabled it will not be possible to enable Link and a tooltip will explain that Link must first be disabled in order to achieve this--and visa versa when Link is enabled instead. 

![Link settings tooltip](https://cdn-std.droplr.net/files/acc_1104206/u6dF9z)
_Visual of the tooltip we will need to add to our settings screen_

Link's short-lived time as an express checkout button is herein terminated, as Stripe does not currently allow us to implement Link as a true express checkout method within the current restraints. This may be possible with Stripe's upcoming new UPE and PRB implementation flows, but for now we are closing this chapter. So shall it be written, so shall it be done.

**Comment by FangedParakeet** [comment]
I'm un-assigning myself from this task for now, in case anyone else would like to take over these changes, since the PR that I had prepared is now mostly redundant. Though, if no one beats me to this issue in time, I'll be back...

**Comment by timur27** [comment]
> We will restore the existing Link implementation to the split UPE branch instead and make sure that it is still compatible with our shortcode and blocks checkouts: this may involve adding a new gateway for Link. Nonetheless when active Link will listen to the email field and trigger automatically as usual, unfettered by WooPay.

As we are iterating on the approach, I'd like to align on the following as part of the refinement before the implementation:

* What are the acceptance criteria for the Stripe Link appearance on the shortcode checkout? Should it be an additional payment gateway on the page together with the Card Element? 
<p align="center"> <img width="327" alt="image" src="https://user-images.githubusercontent.com/22319444/216306890-a7c97300-dc1e-4d76-b2f6-11ca1b1e47e6.png"></p>

Please remember that in this case, after [improving the legacy card element design](https://app.zenhub.com/workspaces/heisenberg-63586f2ec60359001310fa6a/issues/gh/automattic/woocommerce-payments/5273) we'll have two almost identical payment methods on the list on after another, which might be confusing for users.

* Since Link and WooPay are now mutually exclusive, should we try to switch the legacy card with [the Stripe Card Payment element](https://stripe.com/docs/payments/payment-element) and try to integrate it with Link when the latter is enabled and thus WooPay disabled? 

* Is _[Link for the Card Element](https://stripe.com/docs/payments/link-card-element)_ something we considered before, or was it not there when we refined this issue? At first glance, it seems to me that this feature doesn't allow us to use our own email form but instead enter the email after opening the Link modal. Still, maybe there is some opportunity for us to test out here?

* On the Blocks checkout, in case we will proceed with a separate payment method for the Link, we need to consider the current behavior of Blocks - if a user selects Stripe Link and authenticates successfully but then switches over to the legacy card and switches back to the Stripe Link, they'll need to authenticate again. The technical reason behind this is that Blocks are remounting the payment element each time after its being selected again. It looks like this can be improved, but it will require some work directly in the Blocks repository. See p1675070604079049-slack-C02TS23QJ1X for more details.


cc @FangedParakeet @frosso @mdmoore

**Comment by frosso** [comment]
@timur27 thanks for your notes here!
I'm fully in support of making Link compatible with the Card Payment Element again (since WooPay will be disabled) 👍 I wonder how it will affect the user experience, or if merchants would be confused by it (inb4 support ticket stating "my card element looks different when I enable WooPay").
In the longer term, we should be able to get the Card Payment Element to behave correctly with WooPay once we switch the UPE implementation with the "beta" one. So ultimately, we will achieve the same goal.

I'm not sure whether "Link for the Card Element" was considered before 👀 I'll defer to @FangedParakeet for this 👀 

> On the Blocks checkout, in case we will proceed with a separate payment method for the Link, we need to consider the current behavior of Blocks [...]

Thanks for investigating this! I agree, it doesn't seem optimal from the customer's perspective 😬 in the future, would switching the implementation to Link for the Card Element help us, here?

**Comment by timur27** [comment]
> I wonder how it will affect the user experience, or if merchants would be confused by it (inb4 support ticket stating "my card element looks different when I enable WooPay").

Good point. I think that after https://github.com/Automattic/woocommerce-payments/issues/5273 is completed, there will be no difference between the legacy card element and Stripe Card Payment Element.

> I agree, it doesn't seem optimal from the customer's perspective 😬 in the future, would switching the implementation to Link for the Card Element help us, here?

If we use Link in the Stripe Card Element, it will simplify checkout to one payment method (so no problem there because no switching between payment method as such is needed), but users may still face limitations if switching between saved payment methods and the card payment method. 

However, Link behaves this way on the `develop` as well, which makes me feel that probably it was acceptable some time ago.

**Comment by frosso** [comment]
Closing because fixed by https://github.com/Automattic/woocommerce-payments/pull/5515 . 
But the discussion can continue!
